### ISSUE TITLE
iroha_docs_draft

### ISSUE THREAD

**Issue by MizukiSonoko** [hyperledger/iroha#99]
# Sorry... currently, It's  written by Japanese only.
  
- Asset in Iroha  
- Operation
- Architecture   
  
# Asset in Iroha  
  
## Object
irohaで扱うデータの事。種類としてAsset,Domain,Account,Message,Peerがある。  
AssetにはSmartContractを紐つけることができる。
  
### Asset  
最も基本的なObject,文字列をvalueとして持つ。  
#### 要素   
  
- domain (Domain) // 属しているDomain (uuidだけでいい?)
- name  (string)
- value (string)
  
### Currency
Assetのvalueを整数値にしたもの。Use-Caseが多いので分離した。
必要か用検討
#### 要素   
  
- domain (Domain) // 属しているDomain (uuidだけでいい?)
- name  (string)
- value (int)

### Domain
Assetの集合体。  
#### 要素   
  
- ownerPublicKey (string) // 配列の方が良い?
- name (string) 
  
### Account  
irohaを操作できるユーザ。publicKey,name,操作可能なAssetを情報として持つ.  
#### 要素   
  
- ownerPublicKey (string)
- name (string) 
- assets ([Asset]) // domainのとの関係性は?
  
### Peer
sumeragiを構成するpeerの情報を持つ
#### 要素   
  
- publicKey (string)
- ip (string) // IPv4,v6はどう区別する?
- some value  // 信頼度とか
    

### 関係性
```
+--------------+  has  +---------+
|    Domain    |<------| Account |
+--------------+       +---------+
    |       | belong
    v       v
+------+ +------+
|Asset1| |Asset2|
+------+ +------+
    |       |
====================    
    |       |       +-----------------+
    |       └─------| smart contract2 |
    |               +-----------------+
+-----------------+
| smart contract1 |
+-----------------+
```  
イメージとしては現行のURLシステムのようなものです。
それぞれのAssetは特定のDomainに属します。Domainの管理者がAssetを自由に操作することができます。
ex
```
domain:soramitsu 
asset: mizuki
asset: iori

mizuki.soramitsu
iori.soramitsu
toyama.soramitsu
```
  
# Operation
  
### Add
追加をする
#### 要素   

- object  
  
### Transfer 
転送をする
#### 要素   

- receiver
- object  

### Update 
更新をする
#### 要素   
- object  
  
  
### Remove
削除をする
#### 要素   

- object  

### Contract
Smart contractを実行する
#### 要素   

- smart contract name  
- command  
- object  
  
    
### Transaction Command   
```
              |
              |
     ┌--------+-------┐
     |                |
     |       ┌--------|-------┐
     |       | Smart Contract |
     |       └----------------┘                   // Alias まわり(提案)                     // ReadアクセスもTxにする？
     +-----------+-----------+-----------+        
┌----|-----┐┌----|-----┐┌----|-----┐┌----|-----┐  ┌----|-----┐┌----|-----┐┌----|-----┐    ┌----|-----┐┌----|-----┐      
|  Update  ||   Add    || Transfer ||  Remove  |  |  Alias   ||  UnAlias ||  Split   |    |   Find   ||   Get    | 
└----------┘└----------┘└----------┘└----------┘  └----------┘└----------┘└----------┘    └----------┘└----------┘
```

(**WIP** UTXO対応)多分
Transferをうまく使ってownerを移行する形で  
```
Transfer<OwnerOf<Asset>>
```

## Event
irohaのDBを更新するときは基本的にConsensusEventを発行し、Sumeragiに送る。  
// **WIP** transactionを発行したほうが良いのでは  
Sumeragiはそれを合意形成する。

```
+-----------------+
|      Object     | <- Asset, Domain, Account and Peer
+-----------------+ 
|     command     | <- Add, Transfer, Update, Remove and Contract
+-----------------+
|   transaction   | <- sender, timestamp, signatures and hash
+-----------------+
| consensus event | <- sumeragi's signatues, state( Undetermined, Commited)
+-----------------+
```
  
### 発行方法の例(C++)  
```c++
auto event = ConsensusEvent<Transaction<Add<Account>>>(
    issuerPublicKey,   // Transaction 発行者
    accountPublicKey,  
    alias
);
```

```c++
auto event = ConsensusEvent<Transaction<Transfer<Asset>>>(
    issuerPublicKey, // Transaction 発行者
    senderPublicKey, // Asset送り主
    receiverPublicKey,
    assetName,
    value
);
```

コレはC++上での話でしか無いので他のFrontから叩く場合flatbuf(protobuf)を直接叩き皇に合意要求をした方が良いです。

# Architecture   

## Iroha Architecture (currently)

```
 front API
   |(1)
┌-------------┐
+ Cappuccino  +------------------------
└--+----------┘
   |                
   |                                   
┌--┼----------┐         ┌---------------┐
|  V sumeragi |(3)      |  Command exec |
|             ----------+>              |
└--|-------A--┘         └----+----------┘
   |(2)    | valdation       |
   |       └-----------------┼------┐
   |                         |      |
   |consensus building       |(4)   |
┌------------┐             ┌-V------+------┐     ┌-------┐
| connection |             |   repository  |------ model |
└------------┘             └---------------┘     └-------┘
      |                             |
Infra=|=============================|===========
      |implement                    |implement
   ┌------┐                       ┌-----┐
   | grpc |                       | D B |
   └------┘                       └-----┘
```
  
  
## Iroha Architecture (future)

```
 Input (using flat buffer)     Input(http server)
   |(1)                         |
---+----------------+-+---------V----+--
   |----------------┼-┼  Cappuccino  | Cappuccino is option. for http server
   |                | └--------------┘        
┌--┼----------┐     |   
|  | sumeragi |     |   * User programable     
|┌-V---------┐|(2)  | ┌-------------┐   
|| stateless ┼┼-----┼-┼->           |
|├-----------┤|(3)  | | chain code  |  
|| stateful <┼┼-----┼-┼-            |  
|└-|-------|-┘|     | └-┼----┼------┘
└--┼-------┼--┘     └---┼----┼-----------------   
   |       |            |    |
   |       └------------┼----┼------┐
   |                    |    |      |
   |consensus building  |    |      |(4)
┌--┼---------┐          |  ┌-┼------V------┐     ┌-------┐
| connection |          |  |   repository  |------ model |
└-----┼------┘          |  └--------┼------┘     └-------┘
      |                 |           |
Infra=┼=================┼===========┼===========
      |implement        |implement  |implement
   ┌--┼---┐          ┌--┼---┐    ┌--┼--┐
   | grpc |          | JAVA |    | D B |
   └------┘          └------┘    └-----┘
```

**Comment by MizukiSonoko** [comment]
- Domain,Assetにexpireを追加した方が良さそう。
- Domainの所有者は現在1つのアカウントのみ
- Message Objectを追加

**Comment by MizukiSonoko** [comment]
# This is Final draft I think

- Asset in Iroha  
- Operation
- Architecture   
  
# Asset in Iroha  
  
## Object
Object iroha can operate it. There is kind of Currency,  Asset,Domain,Account,Message and Peer. 
Associate SmartContract with Asset.
  
### Asset  
It has the most basic object, "hash map" as value.
#### Elements  
- domainId (Domain) // domain that is attached
- name  (string) 
- value (map<string,Object>)
  
### Currency
It is Asset that it has only integer value.
In this of Virtual Currency, it is used.

####  Elements
- domainId (Domain) domain that is attached
- name  (string)
- value (int)

### Domain
The set of Asset.  
####  Elements 
- ownerPublicKey (List< string > )
- name (string) // length 255
  
### Account  
A user who can operate iroha's data. It has publicKey, name, Operable Asset as information.
#### Elements     
- ownerPublicKey (string) 
- name (string) // length 255
- assets ([Domain])
  
### Peer
Information on the peer making up sumeragi
#### Elements
  
- publicKey (string)
- ip (string) // IPv4,IPv6
- some value  // trust value
    

### Relation
```
+--------------+  has  +---------+
|    Domain    |<------| Account |
+--------------+       +---------+
    |       | belong
    v       v
+------+ +------+
|Asset1| |Asset2|
+------+ +------+
    |       |
====================    
    |       |       +-----------------+
    |       └─------| smart contract2 |
    |               +-----------------+
+-----------------+
| smart contract1 |
+-----------------+
```  

ex
```
domain:hyperledger
asset: mizuki
asset: iori
asset: toyama

mizuki.hyperledger
iori.hyperledger
toyama.hyperledger
```
  
# Operation
I plan to support the following commands:

- Add
- Transfer
- Update
- Remove
- Batch
- Unbatch

# Description
## Add
This does insertion, installation, creation, registration and addition.
EX)
```
Add<Account> = Account registration.
Add<Asset>   = Create asset.
Add<Asset, To<Account>> =  Tie asset to account.

=========
Add< Object >
Add< Object, To> = Tie< Object, To>
```
  
## Transfer
This does moving, migrating, and transferring.
EX)
```
Transfer<Asset>  = Transfer asset value.
Transfer<Account> =  Permissions migration
```
  
## Update
This does updating, and modification.
EX)
```
Update<Account>  = Update account information.
```

## Remove
This does do removing.
EX)
```
Remove<Account>  = Remove account
```

## Batch, Unbatch
This does batching of transactions.
under domain.
EX)
```
Batch<Add<Account>,Transfer<Asset>,Remove<Account>>  = Account registration, Transferring asset and Removing account.
Unbatch = Unbatch
```

# Wrapper
I think Iroha will be support this tokens.

- To

# Architecture   

## Iroha Architecture (currently)

## API layor added iroha

```
 Input (using flat buffer / http(optional)) from client like Android, iOS, Web...
   |(1)                          
   | 
   |
---+------------------------------
   | * User programable (Provide SDK / default grpc api) 
 +-┼-------------------+
 |     API Layor       | 
 +-┼--------------┼----+
---┼--------------┼----------------
   |              |
   |Transaction   |Query
   |   builder    |  repository
┌--┼----------┐   |    
|  | sumeragi |   |    * User programable     
|┌-V---------┐|(2)|   ┌-------------┐   
|| stateless ┼┼---┼---┼->           |    
|├-----------┤|(3)|   | chain code  |     
|| stateful <┼┼---┼---┼-            |     
|└-|-------|-┘|   |   └-┼----┼------┘    
└--┼-------┼--┘   |     |    |
   |       |      └-----┼----┼-----------┐
   |       └------------┼----┼------┐    |
   |                    |    |      |    |
   |consensus building  |    |      |(4) |
┌--┼---------┐          |  ┌-┼------V----v-┐     ┌-------┐
| connection |          |  |   repository  |------ model |
└-----┼------┘          |  └--------┼------┘     └-------┘
      |                 |           |
Infra=┼=================┼===========┼===========
      |implement        |implement  |implement
   ┌--┼---┐          ┌--┼---┐    ┌--┼--┐
   | grpc |          | JAVA |    | D B |
   └------┘          └------┘    └-----┘
```

## SDK sample

### Python Sample 

preview
```python
>> from iroha import Transaction, Repository, Asset
>> 
>> # This is option
>> Transaction.config( 
>>    sumeragi_ip = "localhost", 
>>    sumeragi_port = "51051"
>> )
>> 
>> asset = Asset(name = "Sample", value = "100")
>> tx = Transaction.build(command = "Add", sender = "publicKey", object = asset)
>> Transaction.send(tx) // Send transaction to sumeragi.
>> 
>> sample = Repository.findByName("Sample") // Read data from ledger, at this time
>> print(sample.value)
>> 
```

**Comment by MizukiSonoko** [comment]
ip (string) // IPv4,IPv6 -> use DNS like  alias

**Comment by masterDev1985** [comment]
Some best practices for your Iroha net code:
- Use random backoffs for all connection retries.
- Never use static data to establish connections to your other peers.
- Include retries for every network-related task the peers perform.

**Comment by takemiyamakoto** [comment]
Below is my proposal for the transaction API, adding multisignatory:

# Objects

- Asset
- Domain
- Peer
- SimpleAsset
- Signatory

# Commands

- Add
- Transfer
- Update
- Remove
- Batch
- Unbatch
- 

# Architecture:

```
 Input (using flat buffer / http(optional)) from client like Android, iOS, Web...
   |(1)                          
   | 
   |
---+------------------------------
   | * User programable (Provide SDK / default grpc api) 
 +-┼-------------------+
 |     API Layer       | 
 +-┼--------------┼----+
---┼--------------┼----------------
   |              |
   |Transaction   |Query
   |   builder    |  repository
┌--┼----------┐   |    
|  | sumeragi |   |    * User programable     
|┌-V---------┐|(2)|   ┌-------------┐   
|| stateless ┼┼---┼---┼->           |    
|├-----------┤|(3)|   | chain code  |     
|| stateful <┼┼---┼---┼-            |     
|└-|-------|-┘|   |   └-┼----┼------┘    
└--┼-------┼--┘   |     |    |
   |       |      └-----┼----┼-----------┐
   |       └------------┼----┼------┐    |
   |                    |    |      |    |
   |consensus building  |    |      |(4) |
┌--┼---------┐          |  ┌-┼------V----v-┐     ┌-------┐
| connection |          |  |   repository  |-----| model |
└-----┼------┘          |  └--------┼------┘     └-------┘
      |                 |           |
Infra=┼=================┼===========┼===========
      |implement        |implement  |implement
   ┌--┼---┐          ┌--┼---┐    ┌--┼--┐
   | grpc |          | JAVA |    | D B |
   └------┘          └------┘    └-----┘
```

**Comment by MizukiSonoko** [comment]
In addition, I think this command 

## Tie 
```
Link<Asset>( contractName, assetName); or 
Attach<Asset>( contractName, assetName);
```

## Annotate
```
Annotate<Asset>( assetName, text);
```

**Comment by MizukiSonoko** [comment]
This is my proposal for future asset management system.

## Asset
```
Asset
 ├── Datamodel
 └── Function
```

### Datamodel
It is written by flatbuffer scheme

#### Sample1
```
table Currency{
  name: string (required);
  value: int (required);
}
```

#### Sample2
```
table Message{
  name: string (required);
  message: string (required);
}
```

#### Sample2
```
table Account{
  name: string (required);
  publicKey: string (required);
  belong: string (required);
  permissions: [Permission] (required); 
}
```

### Function
asset's behavior

#### Sample1
```
Currency
 ├── Add
 └── Transfer
```
Add: Issueable
Transfer: remittable


#### Sample1
```
Message
 ├── Add
 ├── Remove
 └── Transfer
```
Add: Issueable
Remove: Removable
Transfer: remittable

# Permission
I think Iroha should have permission function.

## Domain
Domain is like Unix user group
```
┌---------------- Iroha -------------------┐
|              o Root                      |
|┌-- default -┐┌-- Domain1 -┐┌-- Domain2 -┐|
||   o user   ||   o root   ||   o root   ||
||   o user   ||   o user   ||   o user   ||
|| ┌ Asset1 ┐ || ┌ Asset1 ┐ || ┌ Asset1 ┐ ||
|| |        | || |        | || |        | ||
|| └--------┘ || └--------┘ || └--------┘ ||
|└------------┘└------------┘└------------┘|
└------------------------------------------┘
```
Iroha has more than one Iroha-root.
Iroha-root has all parmission for iroha-ledger (Add peer, Add user)

Only Iroha-root can create new domain firstly.
domain has more than one domain-root.
Iroha-root has all parmission for own domain (Add asset, Add user)
basically, user cannot access other domain.

Only Iroha-root can create new user firstly.
User created belongs default domain.

## Action(Iroha root)
- Add/Remove user
- Add/Remove domain
- Add/Remove peer
- Grant user

## Action(Domain root)
- Add/Remove user
- Invite user from other domain
- Grant user
- Add/Remove asset
- Read/Update asset
- Execute smart contract

## Action (Ordinary user)
- Read/Update asset

**Comment by MizukiSonoko** [comment]
Ok, This finished own role
